### PR TITLE
Dx 2184 gk3

### PR DESCRIPTION
**Pull request by suluvai**
# Summary
This is covering all the backwards compatible scenarios, with versions of core-sdk
1. Between `1.0.0-beta.1` and `2.0.0`
2. version `2.0.1`
3. JS/Link SDK

# Why the changes
https://immutable.atlassian.net/browse/DX-2167
https://immutable.atlassian.net/browse/DX-2184

# Things worth calling out
See https://github.com/immutable/imx-core-sdk/pull/341

# Before merging
- [ ] ***For Immutable developers:*** Do the changes in this PR require documentation updates? Please refer to this [SDK documentation guide](https://immutable.atlassian.net/wiki/spaces/PPS/pages/1916994017/SDK+documentation+guide) and list links to documentation update PRs (they don't have to be merged) below (or write `N/A`):
    - [ ] *Add documentation update 1*
    - [ ] *Add documentation update 2*

### CODE REVIEW

File: src/utils/stark/backwards-compatible/crypto.ts [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+/**
+ * Generates a new Stark private key
+ * @returns the private key as a hex string
+ */
+export function generateStarkPrivateKey(): string {
+  const keyPair = starkEc.genKeyPair();
+  // These are/will always be new accounts no need to check.
+  return grindKey(keyPair.getPrivate());
```
Comment (cajames) [inline src/utils/stark/backwards-compatible/crypto.ts:149]: Should this be `grindKeyV1` ?

File: src/utils/stark/backwards-compatible/crypto.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+// Gets the account (stark public key) value of the requested user (ethAddress) for Production environment only.
+export async function getStarkPublicKeyFromImx(
+  ethAddress?: string,
```
Comment (cajames) [inline src/utils/stark/backwards-compatible/crypto.ts:154]: This can't be optional.

File: src/utils/stark/starkCurve.ts [hunk 3]
```diff
@@ -1,23 +1,92 @@
...
+  ethAddress?: string,
+): Promise<string> {
+  // >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
+  // This section is legacy crypto.
+
+  // Same code as legacy.getKeyPairFromPath, required to check if the generated private key
+  // goes above the limit when hashed for first time to identify if we need to do backwards
+  // compatibility check:
+  const privateKeyString = legacy.getPrivateKeyFromPath(seed, path);
+  const starkPrivateKey = legacy.grindKey(privateKeyString);
```
Comment (cajames) [inline src/utils/stark/starkCurve.ts:28]: We need to decide what our "default" stark key grind fn will be

File: src/utils/stark/starkCurve.ts [hunk 4]
```diff
@@ -1,23 +1,92 @@
...
-  const keyPair = starkEc.genKeyPair();
-  return grindKey(keyPair.getPrivate('hex'));
+  const keyPair = legacy.starkEc.genKeyPair();
+  return legacy.grindKey(keyPair.getPrivate('hex'));
+}
+
+async function getKeyFromPath(
+  seed: string,
+  path: string,
+  ethAddress?: string,
```
Comment (cajames) [inline src/utils/stark/starkCurve.ts:19]: Can't be optional

**Review by cajames — APPROVED** [review]

File: src/utils/stark/backwards-compatible/crypto.ts (hunk 1 above)
Comment (suluvai) [reply src/utils/stark/backwards-compatible/crypto.ts:149]: yup, done. Now `grindKeyV1` is the default `grindKey`

File: src/utils/stark/starkCurve.ts [hunk 5]
```diff
@@ -1,23 +1,271 @@
...
+  return key.gte(maxAllowedVal);
+}
+
+// Gets the account (stark public key) value of the requested user (ethAddress) for Production environment only.
+export async function getStarkPublicKeyFromImx(
+  ethAddress: string,
+): Promise<{ starkPublicKey: string; accountNotFound: boolean } | undefined> {
+  try {
+    const imxClient = new ImmutableX(Config.PRODUCTION);
+    // Check if the generated stark public key matches with the existing account value for that user.
```
Comment (igx-C) [inline src/utils/stark/starkCurve.ts:160]: This comment seems misplaced. The line below only queries and returns if it's been registered, but it doesn't match the generated/local key in this function.

Comment (suluvai) [reply src/utils/stark/starkCurve.ts:160]: Yup, you are right. Will fix it.

**Review by igx-C — APPROVED** [review]

**Pull request merged**